take type id into account when calling offsetof on reference types

mcr431 opened the pull request:
Previously all offsets of reference types were off by 4 bytes.
This changes the offsetof implementation for reference types to
add an additional 4 bytes and adds tests that ensure using the
offset on classes and structs result in access to the intended
variable.

see https://github.com/crystal-lang/crystal/issues/7967 for context

*Note this is my first time contributing and I'm not super familiar with the entirety of the codebase. If something should done in a different way, please let me know. This is also my first time working with an LLVM project. I wasn't sure if the type_id is the crystal one which seems to be of type Int32, or something internal to LLVM like this (https://llvm.org/doxygen/classllvm_1_1Metadata.html#a91c7b9c7cf6694f41b9030429b582d26) which is an unsigned int.

asterite commented:
I think this is wrong. It's not always 4 bytes. It depends on padding. For example if you have type_id + string, in 64 bits that's padded to 8 bytes and so the offset will be 8 bytes.

I think you just need to add 1 to the offset in case it's a class, that's all.

bcardiff commented:
Thanks @mcr431 for working on this 💚 .

As pointed by @asterite adding a `+ 1` to de element_index when the type is a class (ie: reference type) is the right approach.

Regarding the specs, I would suggest writing asserts based on the values of the ivars and not teir offsets. That way they are less sensitive to changes in the structure layout.

May I ask, how did you find/pick the #7967? I'm curious.

mcr431 commented:
Made the + 1 to element_index change but haven't addressed spec changes yet. Will update later today.

@bcardiff I picked this issue up because it seemed fairly straight forward, had the community:newcomer tag, and you had left a comment with a starting place in the comments so seemed like a good issue to gently get familiar with the compiler.

mcr431 commented:
Made changes to the spec file. not sure if it was exactly what you were intending. I removed the first four tests since they were essentially testing the same thing as the last four but were relying on implementation details rather than use case. I updated the last four to use ivars instead of hard coded values.

bcardiff commented:
Thanks @mcr431! Happy first-time contribution!

bew commented:
Nice fix! Would have been nice to have a little comment in the code explaining the + 1 though
